src/Data/Nat/Primality.agda

```diff
@@ -8,80 +8,331 @@
 
 module Data.Nat.Primality where
 
-open import Data.Empty using (⊥)
 open import Data.Nat.Base
 open import Data.Nat.Divisibility
 open import Data.Nat.GCD using (module GCD; module Bézout)
 open import Data.Nat.Properties
-open import Data.Product.Base using (∃; _×_; map₂; _,_)
-open import Data.Sum.Base using (_⊎_; inj₁; inj₂)
+open import Data.Product.Base using (_×_; map₂; _,_)
+open import Data.Sum.Base using (_⊎_; inj₁; inj₂; [_,_]′)
 open import Function.Base using (flip; _∘_; _∘′_)
 open import Relation.Nullary.Decidable as Dec
-  using (yes; no; from-yes; ¬?; decidable-stable; _×-dec_; _→-dec_)
+  using (yes; no; from-yes; from-no; ¬?; _×-dec_; _⊎-dec_; _→-dec_; decidable-stable)
 open import Relation.Nullary.Negation using (¬_; contradiction)
-open import Relation.Unary using (Decidable)
+open import Relation.Unary using (Pred; Decidable; IUniversal; Satisfiable; _∩_; _⇒_)
 open import Relation.Binary.PropositionalEquality
-  using (refl; cong)
+  using (_≡_; _≢_; refl; cong)
 
 private
   variable
-    n : ℕ
+    d k m n p : ℕ
+
+  instance
+    nt[2] : NonTrivial 2
+    nt[2] = nonTrivial {0}
 
 ------------------------------------------------------------------------
 -- Definitions
 
+-- Definition of having a non-trivial divisor below a given bound
+
+record HasBoundedNonTrivialDivisor (k n : ℕ) : Set where
+  constructor hasBoundedNonTrivialDivisor
+  field
+    {divisor} : ℕ
+    .{{nt}} : NonTrivial divisor
+    d<k     : divisor < k
+    d∣n     : divisor ∣ n
+
+-- smart constructors
+
+hasBoundedNonTrivialDivisor≢ : .{{NonTrivial d}} → .{{NonZero n}} →
+                    d ≢ n → d ∣ n → HasBoundedNonTrivialDivisor n n
+hasBoundedNonTrivialDivisor≢ d≢n d∣n = hasBoundedNonTrivialDivisor (≤∧≢⇒< (∣⇒≤ d∣n) d≢n) d∣n
+
+hasBoundedNonTrivialDivisor>1 : 1 < d → d < n → d ∣ n → HasBoundedNonTrivialDivisor n n
+hasBoundedNonTrivialDivisor>1 1<d = hasBoundedNonTrivialDivisor
+  where instance _ = n>1⇒nonTrivial 1<d
+
+hasBoundedNonTrivialDivisor∣ : HasBoundedNonTrivialDivisor k m → m ∣ n →
+                               HasBoundedNonTrivialDivisor k n
+hasBoundedNonTrivialDivisor∣ (hasBoundedNonTrivialDivisor d<k d∣m) m∣n
+  = hasBoundedNonTrivialDivisor d<k (∣-trans d∣m m∣n)
+
 -- Definition of compositeness
 
-Composite : ℕ → Set
-Composite 0 = ⊥
-Composite 1 = ⊥
-Composite n = ∃ λ d → 2 ≤ d × d < n × d ∣ n
+Composite : Pred ℕ _
+Composite n = HasBoundedNonTrivialDivisor n n
+
+-- bounded equivalent
+CompositeUpTo : Pred ℕ _
+CompositeUpTo n = ∃⟨ (_< n) ∩ HasNonTrivialDivisor ⟩
```

Comment:
I'm a bit confused about the purpose of this definition? Is it equivalent to `Composite`? Should it be private in this module?


I also have to confess I find this point free style of writing the predicates, (e.g. see `PrimeUpTo` as well) very hard to parse. Would you be against just writing out the full lambda under the exists instead?

Reply:
Both good points. I was on a hobby horse about bounded predicates and PRA...

Regarding 'point-free' predicate definitions, happy to let this go (though I'd always prefer the `∃-syntax` notational style, which this syntax for `Relation.Unary.Satisfiable` affords...). 

Regarding each of the `UpTo` definitions:
* they all follow the same pattern, so it seemed worthwhile emphasising that, even though the instance for `Composite` is ... pointless, inelegant, variation on the definition;
* the `Decidable` property of the public definitions is proved by delegation to the decidability of the `UpTo` versions, and indeed, could be finessed via `Decidable.map` plus an explicit equivalence proof, because they are each equivalent to their `UpTo` counterparts;
* whose proof of `Decidable` can be *automatically* read off the definition by translating `forall` to `all?`, `exists` to `any?`, and so on for the propositional connectives, because the boundedness supplies the requisite bound for those combinators... (a more ambitious refactoring might try to obtain these things by `Reflection`, but that's for another time!);

TL;DR: there's a longer, different, story to tell about these definitions than there is space to develop in this PR... but I had wnated to leave enough breadcrumbs for me (or someone else!?) to be able to follow up on such a story later...

Reply:
Okay thanks for the explanation :+1: